### PR TITLE
feat(query): refactor vacuum temp files

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

feat(query): refactor vacuum temp files

-  `VacuumTempOptions::QueryHook` only vacuum temp files after the query finish
   1. It will vacuum files by hook limit
   2. If there are files remained, store ${query_id}_${index}.list` mark that there are remaining files to be cleaned in the future. 

-  `VacuumTempOptions::VacuumCommand` is called via command `VACUUM TEMPORARY FILES`
    It will list the spill directory
   1. If it scans the meta file, it will clean the data by meta
   3. If it scans the directory, it will get the meta file firstly. If found then delete by meta else delete by list
   4. If it scans other files, delete by iterator.

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [x] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/databendlabs/databend/17089)

### CODE REVIEW

File: src/query/ee/src/storages/fuse/operations/vacuum_temporary_files.rs [hunk 1]
```diff
@@ -34,229 +36,228 @@ const DEFAULT_RETAIN_DURATION: Duration = Duration::from_secs(60 * 60 * 24 * 3);
...
-                        };
-
-                    vacuum_finished_query(
-                        &abort_checker,
-                        start_time,
-                        &mut removed_temp_files,
-                        &mut total_cleaned_size,
-                        &mut batch_size,
-                        &de,
+                let meta = operator.stat(de.path()).await;
```
Comment (Xuanwo) [inline src/query/ee/src/storages/fuse/operations/vacuum_temporary_files.rs:90]: We can use `de.metadata().mode()` to check if it's a dir.

File: src/query/service/src/sessions/query_ctx.rs [hunk 2]
```diff
@@ -387,6 +426,42 @@ impl QueryContext {
...
+            })
+            .cloned()
+            .collect::<Vec<_>>();
+
+        w.clear();
+
+        if !remote_spill_files.is_empty() {
+            let location_prefix = self.query_tenant_spill_prefix();
+            let node_idx = self.get_cluster().ordered_index();
+            let meta_path = format!("{}/{}_{}.meta", location_prefix, self.get_id(), node_idx);
```
Comment (Xuanwo) [inline src/query/service/src/sessions/query_ctx.rs:447]: the `meta`'s name is a bit confuse, how about using `spill_idx` or something?

File: src/query/ee/src/storages/fuse/operations/vacuum_temporary_files.rs (hunk 1 above)
Comment (sundy-li) [reply src/query/ee/src/storages/fuse/operations/vacuum_temporary_files.rs:90]: we need to check `last_modified`, so `stat` op can't be avoided.

Comment (Xuanwo) [reply src/query/ee/src/storages/fuse/operations/vacuum_temporary_files.rs:90]: > we need to check `last_modified`, so `stat` op can't be avoided.

`de.metadata()` will also have `last_modified` for most object storage services.

File: src/query/service/src/spillers/serialize.rs [hunk 3]
```diff
@@ -158,7 +160,7 @@ fn bare_blocks_to_parquet<W: Write + Send>(
     let table_schema = infer_table_schema(&data_schema)?;
 
     let props = WriterProperties::builder()
-        .set_compression(Compression::LZ4_RAW)
+        .set_compression(Compression::SNAPPY)
```
Comment (forsaken628) [inline src/query/service/src/spillers/serialize.rs:163]: How much of a performance difference is there between snappy and lz4?

**Review by forsaken628 — APPROVED** [review]

File: src/query/service/src/spillers/serialize.rs (hunk 3 above)
Comment (sundy-li) [reply src/query/service/src/spillers/serialize.rs]: It's change by mistake, thanks for the tips.

**Review by everpcpc — APPROVED** [review]

**Pull request merged**